### ISSUE TITLE
Pre-Publish Checklist: Use WASM AMP validator

### ISSUE THREAD

**Issue by BrittanyIRL** [GoogleForCreators/web-stories-wp#8477]
While setting up AMP validation checks in the checklist, Pascal mentioned that now the validator is also available in Web Assembly and we could pull it in without needing to update PHP at all. 

For the time being, the PHP is working just fine. This method could be more suitable. This ticket is to explore and potentially implement that and retire the php implementation that was added in #8443. 

Original thread: https://github.com/google/web-stories-wp/pull/6185/files#r675183318

**Comment by swissspidy** [comment]
@BrittanyIRL Turns out we **must** address this very soon.

As per https://github.com/ampproject/amphtml/issues/36067#issuecomment-919532568, the JS version of the validator is being deprecated in favor of the WASM version. It was already removed once but the full removal might be delayed by another few weeks.

So we need to fix this in our next release.

This requires quite some changes on our end:

- The WASM version should ideally be lazy-loaded via a dynamic import, instead of the current way.
  So this needs to be removed:  
  https://github.com/google/web-stories-wp/blob/fc7ffa784766dc49011f897eaa077fc4e31a1a87/includes/Admin/Editor.php#L213-L219
- Error codes aren't strings anymore, but integers.
  A mapping of strings to integers can be found at https://github.com/ampproject/amphtml/blob/cfb102aafc1543788b77f3ef9bd138c753e13097/validator/validator.proto#L853
  `MISSING_URL` is `35`, `INVALID_URL_PROTOCOL` is `37`
  So this needs to be updated here:  
  https://github.com/google/web-stories-wp/blob/fc7ffa784766dc49011f897eaa077fc4e31a1a87/packages/story-editor/src/components/checklist/checks/storyAmpValidationErrors.js#L51-L66
- We might want to load the validator in web worker while we're at it

PLMK if Pea pod has bandwidth for tackling this. If not, Jonny and I might have some room.

**Comment by BrittanyIRL** [comment]
Hey @swissspidy - thanks for the run through. I have meetings all am but I'll get a concrete response back to you by the end of my workday MST.

**Comment by BrittanyIRL** [comment]
@swissspidy  As promised, EOD for me ~

Here's some work in progress: https://github.com/google/web-stories-wp/compare/try/validator 

It uses [`amphtml-validator`](https://amp.dev/documentation/guides-and-tutorials/learn/validation-workflow/validate_amp/#npm-packages-for-ci) combined with [`workerize-loader`](https://github.com/developit/workerize-loader) at the `story-editor` level. I imported inline instead of adjusting the webpack.  is this what you had in mind?

I'm getting yelled at about [unresolving `fs`](https://github.com/ampproject/amphtml/issues/34216), that's why it's kinda just _everywhere_ in this wip. I'm curious if you know how to resolve that (my webpack knowledge is v limited 😝 ).
![Screen Shot 2021-09-15 at 4 49 13 PM](https://user-images.githubusercontent.com/10720454/133527990-46903474-9fd2-4270-a8e8-90a63dea471b.png)

If this is the right direction and you can just lend some eyes on the webpack config/import issues then I think this is almost there and just needs the error codes updates and my wip code cleaned up (both of which I'm capable of).

**Comment by swissspidy** [comment]
Neat! That‘s more than what I expected! :-)

I‘ll take a look tomorrow. In the meantime:

- Have you tried using worker-loader instead? That‘s basically the built-in solution in webpack for web workers.
- Is it possible to load the validator from https://cdn.ampproject.org/v0/validator_wasm.js instead so it‘s evergreen? The npm package was built for a node context, not the web.

**Comment by BrittanyIRL** [comment]
Eh, it's a starting point, right?! 

So `web-worker` is fine with me, thanks for suggesting it. Swapped that out this am. 
Loading the validator from the URL supplied is kinda catching me though with webpack. 

(`await import(/* webpackChunkName: "validator-wasm" */ 'https://cdn.ampproject.org/v0/validator_wasm.js')`)  

Regardless of `lazy` or async/await importing the script results in the error `can't resolve ...validator_wasm` which is because of webpack chunking from what i deduce. Curious how you envision getting around this?
 

How do you want to load the validator if not with npm? Can explore using `require(https)` but that feels like overkill.

**Comment by swissspidy** [comment]
> So web-worker is fine with me, thanks for suggesting it. Swapped that out this am.

NB: the config wasn't fully correct yet in your branch, so the web worker wasn't actually loaded.

> Loading the validator from the URL supplied is kinda catching me though with webpack.

External scripts can be imported using [`importScripts`](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts) in a web worker.

That said, I just took a closer look at your POC branch and tried using `importScripts` there.

It all works just fine, except that apparently the web worker script needs to have the same `Cross-Origin-Embedder-Policy` headers as the editor, otherwise the browser won't load it. And unfortunately we can't really control these headers in that case.

Long story short, using a web worker is unfortunately not feasible for this use case.

So... it boils down to two things we need to do:

1. Update the URL here to `https://cdn.ampproject.org/v0/validator_wasm.js`  
  https://github.com/google/web-stories-wp/blob/fc7ffa784766dc49011f897eaa077fc4e31a1a87/includes/Admin/Editor.php#L213-L219
2. Update error codes from strings to integers as per above message

**Comment by BrittanyIRL** [comment]
ah thank you for spending time on this. i'm on it!

**Comment by BrittanyIRL** [comment]
@swissspidy - 😢  

I think you might need to look deeper into this one after all. Something isn't adding up and I know we're on a time crunch.
Made a draft PR summarizing what's happening and where I'm at. I'll keep looking at it but I'm just not getting much further it feels like. https://github.com/google/web-stories-wp/pull/9062
